mesonbuild/modules/i18n.py

```diff
@@ -14,47 +14,243 @@
+        if not self.preset in PRESET_ARGS:
+            raise coredata.MesonException('i18n: Preset "{}" is not one of the valid options: {}'.format(
+                                          self.preset, list(PRESET_ARGS.keys())))
+
+        self.files = []
+
+    def add_files(self, args, kwargs):
+        if len(args) != 1:
+            raise coredata.MesonException('add_files requires at least one positional argument (file).')
+        preset = self._get_preset(kwargs)
+
+        files = {
+            'files': [ path.join(self.interpreter.subdir, f) for f in args ],
+        }
+
+        if 'language' in kwargs:
+            files['language'] = kwargs.pop('language')
+
+        files['keywords'] = self._get_keywords(kwargs, preset)
+
+        flags = mesonlib.stringlistify(kwargs.pop('flags', []))
+        for f in preset['flags']:
+            if f not in flags:
+                flags.append(f)
+        files['flags'] = flags
+
+        extra_args = mesonlib.stringlistify(kwargs.get('extra_args', []))
+        for a in preset['args']:
+            if a not in extra_args:
+                extra_args.append(a)
+        files['extra_args'] = extra_args
+
+        self.files.append(files)
+
+    def merge_files(self, args, kwargs):
+        if len(args) != 1:
+            raise coredata.MesonException('merge_files requires at least one positional argument (file).')
+        preset = self._get_preset(kwargs)
+
+        language = kwargs.pop('language', 'xml')
+        VALID_LANGUAGES = ('xml', 'desktop')
+        if language not in VALID_LANGUAGES:
+            raise MesonException('i18n: "{}" is not a valid language {}'.format(language, VALID_LANGUAGES))
+
+        datadirs = '--datadirs=' + ':'.join(self.datadirs) if self.datadirs else None
+
+        command = [self.environment.get_build_command(), '--internal', 'msgfmthelper',
+                   '@INPUT@', '@OUTPUT@', language, self.subdir]
+        for k in self._get_keywords(kwargs, preset):
+            command.append('--keyword=' + k)
+        if datadirs:
+            command.append(datadirs)
+
+        suffix = kwargs.pop('suffix', None)
+
+        kwargs['command'] = command
+        ret = []
+        for f in args:
+            kwargs['input'] = f
+            if suffix and f.endswith(suffix):
+                f = f[1:f.index(suffix)]
+            kwargs['output'] = f
+            ret.append(build.CustomTarget(f + '_merge', self.interpreter.subdir, kwargs))
+        return ret
+
+    def process(self, args, kwargs):
+        if len(self.files) == 0:
+            raise coredata.MesonException('process requires files to be added.')
+
+        pkg_arg = '--pkgname=' + self.packagename
+        lang_arg = '--langs=' + '@@'.join(self.languages) if self.languages else None
+        datadirs = '--datadirs=' + ':'.join(self.datadirs) if self.datadirs else None
+        pot_data_file = None
+        extra_args = None
+        preset = self._get_preset(kwargs)
```

Comment:
__[flake8]__
* [F841] local variable 'preset' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->